File: src/Structural_Patterns/Decorator/decorator.rb
# One real-world example of the Decorator design pattern in Ruby is a logging decorator for a web application. This decorator adds logging functionality to an existing class without modifying its original code.

# Here's an example implementation:

ruby
class WebApp
  def initialize
    @routes = {}
  end

  def add_route(path, controller)
    @routes[path] = controller
  end

  def handle_request(path)
    if @routes.key?(path)
      @routes[path].call
    else
      "404 Not Found"
    end
  end
end

class LoggingDecorator
  def initialize(web_app)
    @web_app = web_app
  end

  def add_route(path, controller)
    puts "Adding route: #{path}"
    @web_app.add_route(path, controller)
  end

  def handle_request(path)
    puts "Handling request: #{path}"
    @web_app.handle_request(path)
  end
end


# In this implementation, the `WebApp` class represents a basic web application that stores routes and controllers in a hash and handles incoming requests by calling the appropriate controller.

# The `LoggingDecorator` class wraps the `WebApp` class and implements the same `add_route` and `handle_request` methods. It also adds logging functionality to these methods.

# When the `add_route` method is called on the decorator, it logs a message indicating that a new route has been added and calls the `add_route` method on the underlying web app.

# When the `handle_request` method is called on the decorator, it logs a message indicating that a request is being handled and calls the `handle_request` method on the underlying web app.

# This logging decorator can be used in web applications to add logging functionality to an existing class without modifying its original code. By wrapping the web app in a decorator, developers can easily add new features to the application without introducing bugs or breaking existing functionality.

File: src/Structural_Patterns/Decorator/decorator.py
# One real world example of the Decorator design pattern in Python is the use of decorators in web frameworks such as Flask or Django. Decorators are used to modify the behavior of a function or a class in a non-intrusive way.

# For example, in Flask, the `@app.route` decorator is used to define a route for a web application. This decorator takes a URL pattern as an argument and returns a function that handles the request for that URL.


from flask import Flask

app = Flask(__name__)

@app.route('/')
def index():
    return 'Hello, world!'


# In this example, the `@app.route` decorator is used to define a route for the root URL (`/`). The `index` function is decorated with `@app.route('/')`, which means that it will be called when a request is made to the root URL.

# Another example of the Decorator pattern in Python is the use of decorators to add functionality to a class. For example, the `@property` decorator can be used to define a read-only property on a class.

python
class Person:
    def __init__(self, name):
        self._name = name

    @property
    def name(self):
        return self._name


# In this example, the `@property` decorator is used to define a read-only property `name` on the `Person` class. This allows users of the class to access the `name` attribute without being able to modify it directly.

# Overall, the Decorator design pattern is useful in Python applications for adding functionality to functions or classes in a non-intrusive way. It allows for easy modification and extension of existing code without having to modify the original source code.
